blocks/certifier.go

```diff
@@ -0,0 +1,413 @@
+	if err := codec.Decode(data, &msg); err != nil {
+		logger.With().Error("malformed cert msg", log.Err(err))
+		return errMalformedData
+	}
+	if err := c.validate(ctx, logger, msg); err != nil {
+		return err
+	}
+	if err := c.saveMessage(logger, msg); err != nil {
+		return errInternal
+	}
+	return nil
+}
+
+func (c *Certifier) validate(ctx context.Context, logger log.Log, msg types.CertifyMessage) error {
+	lid := msg.LayerID
+	bid := msg.BlockID
+	if state, err := c.getCertState(logger, lid, bid); err != nil {
+		return err
+	} else if state == expired {
+		return errMsgTooOld
+	} else if state == certified {
+		// should still gossip this msg to peers even when this node has created a certificate
+		return nil
+	}
+
+	logger.WithFields(bid)
+	// extract public key from signature
+	pubkey, err := ed25519.ExtractPublicKey(msg.Bytes(), msg.Signature)
+	if err != nil {
+		return fmt.Errorf("%w: cert msg extract key: %v", errMalformedData, err.Error())
+	}
+	nid := types.BytesToNodeID(pubkey)
+	valid, err := c.oracle.Validate(ctx, msg.LayerID, eligibility.CertifyRound, c.cfg.CommitteeSize, nid, msg.Proof, msg.EligibilityCnt)
+	if err != nil {
+		logger.With().Warning("failed to validate cert msg", log.Err(err))
+		return errInternal
+	}
+	if !valid {
+		logger.With().Warning("invalid cert msg", log.Err(err))
+		return errInvalidCertMsg
+	}
+	return nil
+}
+
+func (c *Certifier) saveMessage(logger log.Log, msg types.CertifyMessage) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	lid := msg.LayerID
+	bid := msg.BlockID
+	if state, err := c.getCertStateLocked(logger, lid, bid); err != nil {
+		return err
+	} else if state != pending { // already certified
+		return nil
+	}
+
+	if _, ok := c.certMsgs[lid]; !ok {
+		logger.Fatal("missing layer in cache")
+	}
+	c.certMsgs[lid].signatures = append(c.certMsgs[lid].signatures, msg)
+
+	if created, err := c.tryGenCertLocked(logger, lid, bid, c.certMsgs[lid].signatures); err != nil {
+		return err
+	} else if created {
+		c.certMsgs[lid].done = true
+	}
+	return nil
+}
+
+func (c *Certifier) tryGenCert() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	cutoff := c.layerClock.GetCurrentLayer().Sub(c.cfg.NumLayersToKeep)
+	for lid, ci := range c.certMsgs {
+		logger := c.logger.WithFields(lid, ci.bid)
+		if lid.Before(cutoff) {
+			delete(c.certMsgs, lid)
+			continue
+		}
+		state, err := c.getCertStateLocked(logger, lid, ci.bid)
+		if err != nil {
+			logger.With().Warning("background: failed to check cert", log.Err(err))
+			continue
+		}
+		if state != pending {
+			continue
+		}
+
+		if created, err := c.tryGenCertLocked(logger, lid, ci.bid, ci.signatures); err != nil {
+			logger.With().Warning("background: failed to generate cert", log.Err(err))
+		} else if created {
+			c.certMsgs[lid].done = true
+		}
+	}
+
+	return nil
+}
+
+func (c *Certifier) tryGenCertLocked(logger log.Log, lid types.LayerID, bid types.BlockID, sigs []types.CertifyMessage) (bool, error) {
+	if len(sigs) < c.cfg.CertifyThreshold {
```

Comment:
Realized need to be counting eligibilities, and not signatures.
I may have made the same mistake in my PR.

Reply:
:+1:
i think eligibility count represents weight unit. and cert needs to be atleast 401 weight units

Reply:
fixed.